docs/DynamicCasting.md

```diff
@@ -0,0 +1,532 @@
+# Dynamic Casting Behavior
+
+* Author: [Tim Kientzle](https://github.com/tbkka)
+* Implementation: [apple/swift#29658](https://github.com/apple/swift/pull/29658)
+
+## Introduction
+
+The Swift language has three casting operators: `is`, `as?`, and `as!`.
+Each one takes an instance on the left-hand side and a type expression on the right-hand side.
+
+* The _cast test operator_ `is` tests whether a particular instance can be converted to a particular destination type.  It returns a boolean result.
+
+* The _conditional cast operator_ `as?` attempts the conversion and returns an `Optional` result: `nil` if the conversion failed, and a non-nil optional containing the result otherwise.
+
+* The _forced cast operator_ `as!` unconditionally performs the casting conversion and returns the result.  If an `as!` expression does not succeed, the implementation may terminate the program.
+
+Note: The static coercion operator `as` serves a different role and it's behavior will not be specifically discussed in this document.
+
+The following invariants relate the three casting operators:
+* Cast test: `x is T == ((x as? T) != nil)`
+* Conditional cast: `(x as? T) == (x is T) ? .some(x as! T) : nil`
+* Forced cast: `x as! T` is equivalent to `(x as? T)!`
```

Comment:
Further, do we guarantee the usual algebraic relationships implied by formal `<:`: reflexivity is here, but do we also guarantee transitivity and anti-symmetry?

Reply:
As explained later, Swift casting operators are not transitive (and cannot easily be made so).

I don't believe anti-symmetry is useful in general.  As discussed below, casting is anti-symmetric for struct and enum types.   It's not if `_ObjectiveCBridgeable` is allowed.  Depending on your definitions, I think protocol types and probably Optional also break it.

Reply:
Protocol types break anti-symmetry:  If you have a protocol `P` with a single conforming struct `S`, then any instance of `S` can be cast to `P` and vice-versa.  However, `S` and `P` instances do not behave identically.  For example, each one can have methods that are not visible on the other.

Reply:
In particular, Swift's casting operations cannot be simply divided into "upcast" and "downcast" cases.